calicoctl.py

```diff
@@ -843,6 +839,135 @@ def bgppeer_show(version):
+        endpoint = client.get_endpoint(hostname, container_id, endpoint_id)
+    except KeyError:
+        print "Container is unknown to calico. Tell calico about the " \
+              "container before adding addresses."
+        sys.exit(1)
+
+    # From here, this method starts having side effects. If something
+    # fails then at least try to leave the system in a clean state.
+    if not client.assign_address(pool, ip):
+        print "IP address is already assigned in pool %s " % pool
+        sys.exit(1)
+
+    try:
+        old_endpoint = copy.deepcopy(endpoint)
+        if address.version == 4:
+            endpoint.ipv4_nets.add(IPNetwork(address))
+        else:
+            endpoint.ipv6_nets.add(IPNetwork(address))
+        client.update_endpoint(hostname, container_id, old_endpoint, endpoint)
+    except (KeyError, ValueError) as e:
+        client.unassign_address(pool, ip)
+        print "Error updating datastore. Aborting."
+        sys.exit(1)
+
+    try:
+        container_pid = info["State"]["Pid"]
+        netns.ensure_namespace_named(container_pid)
+        netns.add_ip_to_interface(container_pid,
+                                  address,
+                                  interface)
+
+    except CalledProcessError:
+        print "Error updating networking in container. Aborting."
+        client.unassign_address(pool, ip)
+        sys.exit(1)
+
+    print "IP %s added to %s" % (ip, container_name)
+
+def container_ip_remove(container_name, ip, version, interface):
+    """
+    Add an IP address to an existing calico networked container.
+
+    :param container_name: The name of the container.
+    :param ip: The IP to add
+    :param version: The IP version
+    :param interface: The name of the interface in the container.
+
+    :return: None
+    """
+    address = check_ip_version(ip, version, IPAddress)
+
+    # The netns manipulations must be done as root.
+    enforce_root()
+
+    pool = get_pool_or_exit(address)
+
+    info = get_container_info_or_exit(container_name)
+    container_id = info["Id"]
+
+    # Check the container is actually running.
+    if not info["State"]["Running"]:
+        print "%s is not currently running." % container_name
+        sys.exit(1)
+
+    # Check that the container is already networked
+    try:
+        endpoint_id = client.get_ep_id_from_cont(hostname, container_id)
+        endpoint = client.get_endpoint(hostname, container_id, endpoint_id)
+    except KeyError:
+        print "Container is unknown to calico. Tell calico about the " \
+              "container before removing addresses."
+        sys.exit(1)
+
+    try:
+        old_endpoint = copy.deepcopy(endpoint)
+        if address.version == 4:
+            endpoint.ipv4_nets.remove(IPNetwork(address))
+        else:
+            endpoint.ipv6_nets.remove(IPNetwork(address))
+        client.update_endpoint(hostname, container_id, old_endpoint, endpoint)
+    except (KeyError, ValueError) as e:
```

Comment:
I think this is slightly too generic in the case where you attempt to remove an IP address that wasn't added to the container in the first place.  In that case, we should have an error message that says the IP address wasn't found on the container.